src/main/java/org/gaul/s3proxy/azureblob/AzureBlobStore.java

```diff
@@ -571,65 +584,436 @@ public void setBlobAccess(String container, String key, BlobAccess access) {
+                        }
+                        md.update(buffer.array(), 0, read);
+                        buffer.limit(read);
+                        sink.next(buffer);
+                        long nextPosition = position + read;
+                        if (nextPosition >= contentLength) {
+                            sink.complete();
+                        }
+                        return nextPosition;
+                    } catch (IOException e) {
+                        sink.error(e);
+                        return position;
+                    }
+                },
+                position -> {
+                    try {
+                        is.close();
+                    } catch (IOException e) {
+                        // Ignore close errors
+                    }
+                }
+            );
+
+            asyncClient.stageBlock(blockId, body, contentLength).block();
+
+            md5Hash = md.digest();
+
+            if (providedMd5 != null) {
+                if (!MessageDigest.isEqual(md5Hash, providedMd5.asBytes())) {
+                    throw new IllegalArgumentException("Content-MD5 mismatch");
+                }
+            }
+
+        } catch (java.security.NoSuchAlgorithmException e) {
+            throw new RuntimeException("MD5 algorithm not available", e);
+        } catch (BlobStorageException bse) {
+            translateAndRethrowException(bse, mpu.containerName(), mpu.blobName());
+            throw new RuntimeException(String.format(
+                    "Failed to upload part %d for blob '%s' in container '%s': %s",
+                    partNumber, mpu.blobName(), mpu.containerName(), bse.getMessage()), bse);
         } catch (IOException ioe) {
-            throw new RuntimeException(ioe);
+            throw new RuntimeException(String.format(
+                    "Failed to upload part %d for blob '%s' in container '%s': %s",
+                    partNumber, mpu.blobName(), mpu.containerName(), ioe.getMessage()), ioe);
         }
-        String eTag = "";  // putBlock does not return ETag
-        Date lastModified = null;  // putBlob does not return Last-Modified
-        return MultipartPart.create(partNumber, length, eTag, lastModified);
+
+        String eTag = BaseEncoding.base16()
+                .lowerCase().encode(md5Hash);
+        Date lastModified = null;
+        return MultipartPart.create(partNumber, contentLength, eTag, lastModified);
+    }
+
+    /**
+     * Creates a BlockBlobAsyncClient with retries disabled for streaming uploads.
+     * This allows us to stream directly from non-markable InputStreams without
+     * needing temp files or buffering. The S3 client can retry the entire part
+     * upload if needed.
+     */
+    private BlockBlobAsyncClient createNonRetryingBlockBlobAsyncClient(
+            String container, String blobName) {
+        var cred = creds.get();
+        var retryOptions = new RequestRetryOptions(
```

Comment:
Could you hoist this into a static final and use the options for both the sync and async clients?